javaslang/src/main/java/javaslang/collection/LeafType.java

```diff
@@ -0,0 +1,258 @@
+        return (i < 0) ? wrapper
+                       : PRIMITIVES[i];
+    }
+
+    private static int primitiveIndex(Class<?> wrapper) { /* linear search is faster than binary search here */
+        for (int j = 0; j < WRAPPERS.length; j++) {
+            if (wrapper == WRAPPERS[j]) {
+                return j;
+            }
+        }
+        return -1;
+    }
+}
+
+final class ObjectArray extends LeafType<Object> {
+    private static final long serialVersionUID = 1L;
+    static final ObjectArray OBJECT_TYPE = new ObjectArray();
+
+    @Override
+    boolean isPrimitive() { return false; }
+
+    @Override
+    Class<Object> type() { return Object.class; }
+
+    static final Object[] EMPTY = new Object[0];
+
+    @Override
+    Object[] empty() { return EMPTY; }
+
+    private static Object[] cast(Object array) { return (Object[]) array; }
+
+    @Override
+    int lengthOf(Object array) { return cast(array).length; }
+
+    @Override
+    Object get(Object array, int index) { return cast(array)[index]; }
+
+    @Override
+    void setUnsafe(Object array, int index, Object value) { cast(array)[index] = value; }
+
+    @Override
+    Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size) {
+        if (size == 0) {
+            return new Object[arraySize];
+        } else {
+            final Object[] result = new Object[arraySize];
+            System.arraycopy(array, sourceFrom, result, destinationFrom, size);
+            return result;
+        }
+    }
+}
+
+final class IntArray extends LeafType<Integer> {
+    private static final long serialVersionUID = 1L;
+    static final IntArray INT = new IntArray();
+
+    @Override
+    Class<Integer> type() { return int.class; }
+
+    private static final int[] EMPTY = new int[0];
+
+    @Override
+    int[] empty() { return EMPTY; }
+
+    private static int[] cast(Object array) { return (int[]) array; }
+
+    @Override
+    int lengthOf(Object array) { return cast(array).length; }
+
+    @Override
+    Integer get(Object array, int index) { return cast(array)[index]; }
+
+    @Override
+    void setUnsafe(Object array, int index, Object value) { cast(array)[index] = (Integer) value; }
+
+    @Override
+    Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size) {
+        if (size == 0) {
+            return new int[arraySize];
+        } else {
+            final int[] result = new int[arraySize];
+            System.arraycopy(array, sourceFrom, result, destinationFrom, size);
+            return result;
+        }
+    }
+}
+
+@SuppressWarnings("unchecked")
+final class PrimitiveArray<T> extends LeafType<T> {
+    private static final long serialVersionUID = 1L;
+    final Class<T> type;
+
+    PrimitiveArray(Class<T> type) { this.type = type; }
+
+    @Override
+    Class<T> type() { return type; }
+
+    @Override
+    Object empty() { return java.lang.reflect.Array.newInstance(type, 0); }
```

Comment:
> This is a backup plan (i.e. we could spell out every primitive), though I'm not even sure whether this would work/is necessary in Javascipt.

It seems you should create a lot of implementations like `IntArray`?

> what are your thoughts on the rest of the PRs?

I'm very sorry, I can't find a few **consecutive** hours for review your previous PR :(
I still doubt in the effectiveness of tail and head arrays in the `Vector` (because of significantly code complication), but benchmarks are more than convincing :)